pgxn/neon/pagestore_smgr.c

```diff
@@ -235,7 +236,9 @@ typedef struct PrefetchState
 								 * also unused */
 
 	/* the buffers */
-	prfh_hash  *prf_hash;
+	prfh_hash	*prf_hash;
+	int			max_shard_no;
+	uint8		shard_bitmap[(MAX_SHARDS + 7)/8];
```

Comment:
What is `shard_bitmap`?

Reply:
We have a number of buffered prefetch requests for different shards. Before waiting for response we need two flush all pending requests. And to do it we need to know which shards are effected. This is why I am using this bitmap. Correspondent bit in bitmap is set when we send some request to the shards and cleared by flush.

Reply:
> Is there a single prefetch ring for all shards? The responses might come in different order across the pageservers, so I would assume there to be a separate prefetch ring for each shard. Connections to different pageservers will also be lost and re-established at different times.

No there is still the single prefetch ring for all shards. But each prefetch slot has shard number. When we are waiting for this slot, we wait response from the particular shard. @jcsp said that we can have up to hundred of shards in future. Having separate ring for each shard in this case will be very inefficient.